File: src/56.merge-intervals.java
import java.util.ArrayList;
import java.util.List;

/*
 * @lc app=leetcode id=56 lang=java
 *
 * [56] Merge Intervals
 */

// @lc code=start
class Solution {
    public int[][] merge(int[][] intervals) {
        List<int[]> result = new ArrayList<>();
        if (intervals.length == 0) {
            return result.toArray(new int[][] {});
        }
        result.add(intervals[0]);
        List<int[]> subResult;
        for (int i = 1; i < intervals.length; i++) {
            subResult = new ArrayList<>();
            for (int[] value : result) {
                if (intervals[i][0] <= value[1] && intervals[i][1] >= value[0]) {
                    if (intervals[i][0] < value[0]) {
                        value[0] = intervals[i][0];
                    }
                    if (intervals[i][1] > value[1]) {
                        value[1] = intervals[i][1];
                    }
                    intervals[i] = value;
                } else {
                    subResult.add(value);
                }
            }
            subResult.add(intervals[i]);
            result = subResult;
        }

        return result.toArray(new int[][] {});
    }
}
// @lc code=end

File: src/56.merge-intervals.go
/*
 * @lc app=leetcode id=56 lang=golang
 *
 * [56] Merge Intervals
 */

// @lc code=start
func merge(intervals [][]int) [][]int {
	result := make([][]int, 0)
	if (len(intervals) == 0) {
		return result
	}
	// 验证了猜想，不是按序排列的
	// 还需要合并多个interval
	result = append(result, intervals[0])
	var subResult = make([][]int, 0)
	for i, v := range intervals {
		if i == 0 {
			continue
		}
		subResult = make([][]int, 0)
		for j, in := range result {
			if v[0] <= in[1] && v[1] >= in[0] {
				if v[0] < in[0] {
					in[0] = v[0]
				}
				if v[1] > in[1] {
					in[1] = v[1]
				}
				v[0] = in[0]
				v[1] = in[1]
			} else {
				subResult = append(subResult, in)
			}
			if j == len(result) - 1 {
				subResult = append(subResult, v)
			}
		}
		result = subResult
	}
	return result
}
// @lc code=end

